src/mscorlib/shared/System/HashCode.cs

```diff
@@ -0,0 +1,433 @@
+
+            // To see what's really going on here, have a look at the Combine methods.
+            
+            var val = (uint)value;
+            uint position = _length % 4;
+
+            // Switch can't be inlined.
+
+            if (position == 0)
+                _queue1 = val;
+            else if (position == 1)
+                _queue2 = val;
+            else if (position == 2)
+                _queue3 = val;
+            else // == 3
+            {
+                // length smaller than 4?
+                if ((_length ^ position) == 0)
+                    Initialize(out _v1, out _v2, out _v3, out _v4);
+
+                _v1 = Round(_v1, _queue1);
+                _v2 = Round(_v2, _queue2);
+                _v3 = Round(_v3, _queue3);
+                _v4 = Round(_v4, val);
+            }
+
+            // Throw for more than uint.MaxValue fields.
+            _length = checked(_length + 1);
+        }
+
+        public int ToHashCode()
+        {
+            uint position = _length % 4;
+
+            // If the length is less than 3, _v1 -> _v4 don't contain
+            // anything yet. xxHash32 treats this differently.
+
+            uint hash = (_length ^ position) == 0
+                ? MixEmptyState()
+                : MixState(_v1, _v2, _v3, _v4);
+
+            // Multiply by 4 because we've been counting in ints, not
+            // bytes.
+
+            hash += _length * 4;
+
+            // Mix what remains in the queue
+            // Switch can't be inlined right now, so use as few
+            // branches as possible instead.
+
+            if (position > 0)
+            {
+                hash = QueueRound(hash, _queue1);
+                if (position > 1)
+                {
+                    hash = QueueRound(hash, _queue2);
+                    if (position > 2)
+                        hash = QueueRound(hash, _queue3);
+                }
+            }
+
+            hash = MixFinal(hash);
+            return (int)hash;
+        }
+
+#       pragma warning disable 0809
+        // Obsolete member 'memberA' overrides non-obsolete member 'memberB'. 
+        // Disallowing GetHashCode is by design
+
+        // * We decided to not override GetHashCode() to produce the hash code 
+        //   as this would be weird, both naming-wise as well as from a behavioral 
+        //   standpoint (GetHashCode() should return the object's hash code, not 
+        //   the one being computed).
+
+        // * Even though ToHashCode() can be called safely multiple times on this
+        //   implementation, it is not part of the contract. If the implementation
+        //   has to change in the future we don't want to worry about people who
+        //   might have incorrectly used this type.
+
+        [Obsolete("Use ToHashCode to retrieve the computed hash code.", error: true)]
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public override int GetHashCode() => throw new NotSupportedException();
```

Comment:
It might be helpful to include a message with the `NotSupportedException` similar to (or the same as) the obsolete message, added to https://github.com/dotnet/coreclr/blob/master/src/mscorlib/Resources/Strings.resx